### ISSUE TITLE
dart2native performance issue

### ISSUE THREAD

**Issue by jkleiser** [dart-lang/sdk#39260]
I used dart2native to compile a little Dart program that reads through a big word vector file (typically > 1 GB) and prints some statistics. The execution times were a bit disappointing. The compiled native used real 0m37.340s (user 0m35.359s), while the old non-compiled used only real 0m23.449s (user 0m23.894s). This was on a MacBook Air.
My Dart program, vec-test.dart, is listed at the bottom. I run it like this: `dart vec-test.dart some_file.vec`
A vec file (text) may be found at https://fasttext.cc/docs/en/crawl-vectors.html, e.g. https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.no.300.vec.gz

* Dart SDK Version (`dart --version`)
Dart VM version: 2.6.0 (Thu Oct 24 17:52:22 2019 +0200) on "macos_x64"

* Whether you are using Windows, MacOSX, or Linux (if applicable)
macOS 10.14.6, Mojave

```
import "dart:io";
import "dart:convert";
import "dart:async";
import "dart:math";

RegExp norwRx = new RegExp(r"^[a-zæøåé]+\-?[a-zæøåé]*$");

bool nordic(String word) {
  return true;
}

main(List<String> arguments) {
  final filename = arguments.first;
  final file = new File(filename);
  Stream<List<int>> inputStream = file.openRead();
  
  final verbose = arguments.length > 1;
  int wordLines;
  int dims;
  int goodCount = 0;
  double maxLen = 0.0;
  String maxWord;
  
  inputStream
    .transform(utf8.decoder)
    .transform(new LineSplitter())
    .listen((String line) {
        List<String> parts = line.split(" ");
        if (parts.length <= 3) {  // may be a trailing space
          wordLines = int.parse(parts[0]);
          dims = int.parse(parts[1]);
          print("$wordLines word lines, $dims dimensions");
        } else if(norwRx.hasMatch(parts[0])) {
          double sumOfSquares = 0.0;
          for (int i=1; i<=dims; i++) {
            double d = double.parse(parts[i]);
            sumOfSquares += d * d;
          }
          double vLen = sqrt(sumOfSquares);
          goodCount++;
          if (vLen > maxLen) {
            maxLen = vLen;
            maxWord = parts[0];
          }
          if (verbose) {
            print("${parts[0]}: $vLen");
          }
        }
      },
      onDone: () {
        print("\naccepted words: $goodCount");
        print("maxLen=$maxLen, for '$maxWord'");
      },
      onError: (e) { print(e.toString()); }
    );
}
```

**Comment by jamesderlin** [comment]
`RegExp` seems to be much slower with dart2native.  Here is another case:

```dart
#!/usr/bin/env dart

void main() {
  final re = RegExp(r'^/foo/bar/baz/(.+)$');
  final s = '/foo/bar/baz/the_five_boxing_wizards/jump/quickly';

  final results = List<String>(100000);
  final sw = Stopwatch()..start();
  for (var i = 0; i < results.length; i += 1) {
    results[i] = re.firstMatch(s)?.group(1);
  }
  print('${sw.elapsedMilliseconds}ms');
}
```
Running this with directly (i.e., through `dart`) on my Linux laptop, this completes in ~30ms.

Using dart2native and running the result, this instead completes in ~500ms(!).

Dart VM version: 2.6.1 (Mon Nov 11 13:12:24 2019 +0100) on "linux_x64"

**Comment by gnawf** [comment]
I'm running into the same issue on Flutter, where in the release build (which is AOT – [see](https://flutter.dev/docs/perf/rendering/ui-performance)) the `RegExp` performance is heavily degraded – see profiling below. Parsing the same content took 1476ms (AOT) vs 84ms (JIT).

From the profiling, it seems like the culprit is `_ExecuteMatchSticky` – in other profiling I did the culprit is much more exaggerated, in hindsight I should have taken screenshots of those.

<img width="2048" alt="Screen Shot 2020-04-04 at 2 49 01 pm" src="https://user-images.githubusercontent.com/9077461/78416856-c884d600-7688-11ea-93ff-648815a9de15.png">

<img width="2048" alt="Screen Shot 2020-04-04 at 2 48 36 pm" src="https://user-images.githubusercontent.com/9077461/78416855-c3278b80-7688-11ea-8044-2f973bc59b04.png">

My `flutter doctor` output – for Dart version.

```
[✓] Flutter (Channel unknown, v1.15.9, on Mac OS X 10.15.4 19E266, locale en-AU)
    • Flutter version 1.15.9 at /Users/fwang/Documents/Personal/flutter
    • Framework revision cc52a903a8 (4 weeks ago), 2020-03-04 18:59:18 -0800
    • Engine revision 810727bf3f
    • Dart version 2.8.0 (build 2.8.0-dev.11.0 57462f9ca5)
```

Am happy to provide more details if needed.

**Comment by priezz** [comment]
I have the other very simple example, where compiled performance is about 85% worse, than in VM:
```dart
void main() {
  final List<int> result = List(3000);
  for (int i = 0; i < 15; i++) {
    final start = DateTime.now();
    for (int j = 0; j < 3000; j++) {
      result[j] = i;
    }
    int sum = 0;
    for (int j = 0; j < 3000; j++) {
      sum += result[j];
    }
    // final int sum = result.reduce((a, b) => a + b);
    final end = DateTime.now();
    print('${end.difference(start).inMicroseconds}, $sum');
  }
}
```

VM results (you can see heating up in the first 3 rows, that's fine):
```
93, 0
154, 3000
85, 6000
7, 9000
6, 12000
6, 15000
7, 18000
6, 21000
6, 24000
6, 27000
6, 30000
6, 33000
6, 36000
7, 39000
6, 42000
```

dart2native results:
```
13, 3000
12, 6000
11, 9000
11, 12000
12, 15000
11, 18000
11, 21000
11, 24000
11, 27000
11, 30000
11, 33000
11, 36000
11, 39000
13, 42000
```

P.S. When using `final int sum = result.reduce((a, b) => a + b);` to calculate sum instead of the manual loop it takes 2-8 times more time of VM and 2 times more when compiled.

**Comment by jamesderlin** [comment]
I tried my example from https://github.com/dart-lang/sdk/issues/39260#issuecomment-557232973 again with Dart 2.12.1 on the same Linux system.  Running it through `dart` still takes ~30ms, but running the native version (compiled via `dart compile exe`) now takes ~70ms.  It's still worse, but it's significantly improved from the 500ms it took before.

**Comment by mit-mit** [comment]
@mkustermann does this look like a dupe of https://github.com/dart-lang/sdk/issues/37785 ?

**Comment by mkustermann** [comment]
@priezz  Running your example from https://github.com/dart-lang/sdk/issues/39260#issuecomment-611903349 with newest Dart JIT and Dart AOT seems to yield a similar result numbers. Please consider opening a new issue if you see other significant performance issues.

**Comment by mkustermann** [comment]
@gnawf  @jamesderlin `RegExp`s in AOT are significantly slower because they are interpreted in AOT (vs. compiled to native code in JIT). This is due to a trade-off of performance and code size. This is tracked at https://github.com/dart-lang/sdk/issues/37785

**Comment by mit-mit** [comment]
Closing for now

**Comment by mkustermann** [comment]
@jkleiser Running your original example from https://github.com/dart-lang/sdk/issues/39260#issue-518436427 on the vector file mentioned I get this in JIT:
```
%  measure out/ReleaseX64/dart --no-sound-null-safety test.dart cc.no.300.vec
2000000 word lines, 300 dimensions

accepted words: 808856
maxLen=8.12742444689583, for 'iw'
        Memory: 182048 kB  -    1:02.59 real,   66.35 user,     2.65 sys
```
and this in AOT (same as `dart2native` as well as `dart compile exe`):
```
% measure pkg/vm/tool/dart_precompiled_runtime2 --no-sound-null-safety test.dart.aot cc.no.300.vec 
2000000 word lines, 300 dimensions

accepted words: 808856
maxLen=8.12742444689583, for 'iw'
        Memory: 54724 kB  -     1:17.05 real,   80.97 user,     2.69 sys
```

The AOT profile looks like this:
```
+   17.65%  DartWorker       test.dart.aot             [.] Precompiled__LineSplitterSink_9003594__addLines_9003594_716                                                                                                               ▒
+   10.98%  DartWorker       test.dart.aot             [.] Precompiled__Utf8Decoder_9003594_decode8_688                                                                                                                              ◆
+   10.91%  DartWorker       test.dart.aot             [.] Precompiled_double__tryParseDouble_0150898_215                                                                                                                            ▒
+    8.72%  DartWorker       dart_precompiled_runtime  [.] dart::Object::Allocate                                                                                                                                                    ▒
+    6.79%  DartWorker       test.dart.aot             [.] Precompiled__Utf8Decoder_9003594_decode16_686                                                                                                                             ▒
+    5.74%  DartWorker       dart_precompiled_runtime  [.] dart::BootstrapNatives::DN_OneByteString_splitWithCharCode                                                                                                                ▒
+    5.71%  DartWorker       test.dart.aot             [.] Precompiled__OneByteString_0150898_codeUnitAt_3560                                                                                                                        ▒
+    5.48%  DartWorker       dart_precompiled_runtime  [.] dart::UntaggedObject::StoreCompressedArrayPointer<dart::AbstractTypePtr, dart::AbstractTypePtr, (std::__2::memory_order)0>                                                ▒
+    3.03%  DartWorker       dart_precompiled_runtime  [.] dart::OneByteString::SubStringUnchecked                                                                                                                                   ▒
+    2.93%  DartWorker       dart_precompiled_runtime  [.] dart::GrowableObjectArray::Add                                                                                                                                            ▒
+    2.46%  DartWorker       test.dart.aot             [.] Precompiled__OneByteString_0150898__substringUncheckedNative_0150898_3526                                                                                                 ▒
+    2.00%  DartWorker       dart_precompiled_runtime  [.] dart::Heap::AllocateNew                                                                                                                                                   ▒
+    1.85%  DartWorker       test.dart.aot             [.] Precompiled__TwoByteString_0150898_codeUnitAt_3559                                                                                                                        ▒
+    1.62%  DartWorker       dart_precompiled_runtime  [.] dart::String::SubString                                                                                                                                                   ▒
+    1.38%  DartWorker       dart_precompiled_runtime  [.] dart::Array::Grow                                                                                                                                                         ▒
+    1.37%  DartWorker       test.dart.aot             [.] Precompiled____main__anonymous_closure__3678                                                                                                                              ▒
+    0.92%  DartWorker       test.dart.aot             [.] Precompiled__Double_0150898__Double_0150898_fromInteger_217                                                                                                               ▒
+    0.88%  DartWorker       libc-2.31.so              [.] __memmove_avx_unaligned_erms                                                                                                                                              ▒
+    0.87%  DartWorker       test.dart.aot             [.] Precompiled_double__parse_0150898_213                                                                                                                                     ▒
+    0.66%  DartWorker       dart_precompiled_runtime  [.] dart::String::Copy                                                                                                                                                        ▒
+    0.61%  DartWorker       test.dart.aot             [.] Precompiled__StringBase_0150898__lastNonWhitespace_0150898_1970                                                                                                           ▒
+    0.51%  DartWorker       libc-2.31.so              [.] __memset_avx2_erms                                                                                                                                                        ▒
+    0.50%  DartWorker       test.dart.aot             [.] Precompiled__StringBase_0150898__firstNonWhitespace_0150898_1971                                                                                                          ▒
+    0.45%  DartWorker       test.dart.aot             [.] Precompiled__Utf8Decoder_9003594_convertChunked_685                                                                                                                       ▒
+    0.35%  DartWorker       dart_precompiled_runtime  [.] dart::IrregexpInterpreter::Match                                                                                                                                          ▒
+    0.29%  DartWorker       dart_precompiled_runtime  [.] dart::VMHandles::AllocateHandle        
```

There's a few differences between the AOT and JIT profile:

  * Calls to any methods on `String` can be inlined in the JIT (because it knows whether `OneByteString` or `TwoByteString` objects are actually used at runtime). The AOT doesn't know this so it will make virtual calls on `String` (e.g. this call `String.codeUnitAt()` call in the [line_splitter.dart](https://github.com/dart-lang/sdk/blob/00d39e0f6ca057ca81a710b13f459154031bb9dc/sdk/lib/convert/line_splitter.dart#L143))
  * AOT does less aggressive inlining
  * JIT uses compiled RegExp vs AOT uses interpreted (this is tracked at #37785). Though seems minor on the profiler.

Then for both JIT and AOT there's an issue that `String.split(<pattern>)` has a special case for `OneByteStrings` with single character patterns, see [string_patch.dart](https://github.com/dart-lang/sdk/blob/00d39e0f6ca057ca81a710b13f459154031bb9dc/sdk/lib/_internal/vm/lib/string_patch.dart#L998):
```
  List<String> _splitWithCharCode(int charCode)  native "OneByteString_splitWithCharCode";

  List<String> split(Pattern pattern) {
    ...
    if (ClassID.getID(pattern) == ClassID.cidOneByteString) {
      final String patternAsString = unsafeCast<String>(pattern);
      if (patternAsString.length == 1) {
        return _splitWithCharCode(patternAsString.codeUnitAt(0));
      }
    }
    return super.split(pattern);
  }
```

This can cause a big slowdown if one has to go to native (C++ code) for every `split` call.

When fixing this, the resulting profile looks like this in AOT:
```
+   18.45%  DartWorker       test.dart.aot             [.] Precompiled__LineSplitterSink_9003594__addLines_9003594_716                                                                                                               ▒
+   16.75%  DartWorker       test.dart.aot             [.] Precompiled__OneByteString_0150898__splitWithCharCode_0150898_3076                                                                                                        ▒
+   12.42%  DartWorker       test.dart.aot             [.] Precompiled__Utf8Decoder_9003594_decode8_688                                                                                                                              ◆
+   11.95%  DartWorker       test.dart.aot             [.] Precompiled_double__tryParseDouble_0150898_215                                                                                                                            ▒
+    9.08%  DartWorker       test.dart.aot             [.] Precompiled__OneByteString_0150898__substringUncheckedNative_0150898_3505                                                                                                 ▒
+    6.33%  DartWorker       test.dart.aot             [.] Precompiled__OneByteString_0150898_codeUnitAt_3539                                                                                                                        ▒
+    5.08%  DartWorker       test.dart.aot             [.] Precompiled__Utf8Decoder_9003594_decode16_686                                                                                                                             ▒
+    3.95%  DartWorker       test.dart.aot             [.] Precompiled__GrowableList_0150898__grow_0150898_454                                                                                                                       ▒
+    1.49%  DartWorker       test.dart.aot             [.] Precompiled_Stub__iso_stub_AllocateArrayStub                                                                                                                              ▒
+    1.35%  DartWorker       test.dart.aot             [.] Precompiled__TwoByteString_0150898_codeUnitAt_3538                                                                                                                        ▒
+    1.34%  DartWorker       test.dart.aot             [.] Precompiled____main__anonymous_closure__3657                                                                                                                              ▒
+    1.22%  DartWorker       dart_precompiled_runtime  [.] dart::String::SubString                                                                                                                                                   ▒
+    1.01%  DartWorker       test.dart.aot             [.] Precompiled__Double_0150898__Double_0150898_fromInteger_217                                                                                                               ▒
+    0.97%  DartWorker       dart_precompiled_runtime  [.] dart::Object::Allocate                                                                                                                                                    ▒
+    0.96%  DartWorker       test.dart.aot             [.] Precompiled_double__parse_0150898_213                                                                                                                                     ▒
+    0.69%  DartWorker       test.dart.aot             [.] Precompiled__StringBase_0150898__lastNonWhitespace_0150898_1970                                 
```

Basically it spends the majority of code in (decreasing order - i.e. first is most expensive):

  * splitting input into lines: making many small `String` objects for each line
  * splitting line into parts: making many small `String` objects for each part of the line
  * utf8 decoding
  * double parsing

Abstractions (such as using line splitter) are not zero-cost (i.e. compiler cannot always get rid of them).

An optimized implementation for reading such word vector files could do the following instead:

  * Use `dart:ffi` to memory-map the file into the virtual address space (that would avoid copying when reading the file)
  * Finding the line ends manually by walking the bytes
  * utf-8 decoding the lines

**Comment by mkustermann** [comment]
For the slow implementation of `String.split()` I've filed https://github.com/dart-lang/sdk/issues/46352
